[dynamo] Better error message for local_map subgraph mismatches number of inputs/outputs with placement info

xmfan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #164437
* #164433
* #164431
* #163602
* #164340
* #164420
* #164419
* __->__ #164321
* #164296

Reviewed GPT5 summary:

**Summary / Goal**  
Improve error reporting when local_map subgraph input/output counts mismatch placement info.  

**Details**  
- Adds descriptive runtime error messages.  

**Motivation**  
Helps debug local_map misalignments.  

```python
AssertionError: Expecting 2 inputs to local_map function based on placements, but found 1. If the count matches for eager, Dynamo may have flattened inputs to the function or found additional tensors used via closures. Please adjust the input placements to match what the traced graph sees: 
class GraphModule(torch.nn.Module):
    def forward(self, l_args_0_: "f32[8, 8, 16]"):
         # File: /home/xmfan/core/a/pytorch/test/higher_order_ops/test_local_map.py:523 in mismatch_input, code: return x + scalar, scalar
        child: "f32[8, 8, 16]" = l_args_0_ + 10;  l_args_0_ = None
        return (child,)
        .
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela @mlazos

xmfan committed:
[dynamo] Better error message for local_map subgraph mismatches number of inputs/outputs with placement info

[ghstack-poisoned]

xmfan committed:
Update on "[dynamo] Better error message for local_map subgraph mismatches number of inputs/outputs with placement info"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

bdhirsh commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -3588,6 +3588,33 @@ def _call_function(
+        expected_num_inputs = len(in_placements.value)
+        actual_num_inputs = len(body_gmod.graph.find_nodes(op="placeholder"))
+        expected_num_outputs = len(out_placements.value)
+        assert len(body_gmod.graph.find_nodes(op="output")) == 1
```
just curious, was there a bug at some point that would have failed this assert? (I feel like this invariant has been around for a long time)

ezyang reviewed (changes requested):
flagrantly inefficient

ezyang reviewed (approved):
as long as you know about fixing the print_readable i'm going to preapprove so you're not blocked for landing

xmfan committed:
Update on "[dynamo] Better error message for local_map subgraph mismatches number of inputs/outputs with placement info"


Reviewed GPT5 summary:

**Summary / Goal**  
Improve error reporting when local_map subgraph input/output counts mismatch placement info.  

**Details**  
- Adds descriptive runtime error messages.  

**Motivation**  
Helps debug local_map misalignments.  

```python
AssertionError: Expecting 2 inputs to local_map function based on placements, but found 1. If the count matches for eager, Dynamo may have flattened inputs to the function or found additional tensors used via closures. Please adjust the input placements to match what the traced graph sees: 
class GraphModule(torch.nn.Module):
    def forward(self, l_args_0_: "f32[8, 8, 16]"):
         # File: /home/xmfan/core/a/pytorch/test/higher_order_ops/test_local_map.py:523 in mismatch_input, code: return x + scalar, scalar
        child: "f32[8, 8, 16]" = l_args_0_ + 10;  l_args_0_ = None
        return (child,)
        .
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela mlazos

[ghstack-poisoned]
